### PR TITLE
Fix print rules

### PR DESCRIPTION
**Pull request by rgrinberg**
The first few commits are just some refactorings. The last commit fixes `$ dune rules`.

### CODE REVIEW

**Comment by diml** [comment]
IIUC, this PR changes the output of `dune rules -m` so that the dependencies of the action are taken into account. Is that correct?

That doesn't seem right to me, the action dependencies are only needed to compute the final (deps, action, targets) triplet. They shouldn't appear in the `Makefile`  given that these files won't be read by `action`.

**Comment by rgrinberg** [comment]
> That doesn't seem right to me, the action dependencies are only needed to compute the final (deps, action, targets) triplet. They shouldn't appear in the Makefile given that these files won't be read by action.

I might be misunderstanding, but aren't you referring to `Static_deps.rule_deps` here? I.e. things like 	reading the ocamldep file. If so, then those are indeed omitted.

I'm not very familiar with this part of the code is a little more explanation, or if you'd like to push a commit to show what you mean. Both options are welcome.

**Comment by diml** [comment]
Ah never mind, I now see what was the issue with the original code. This code has become quite complicated and is hard to follow. I pushed a few commits to refactor build_system.ml, the new code should be easier to read.

**Comment by rgrinberg** [comment]
Looks good @diml . I rebased the PR and tested things. Why are there changes to the tests?

**Comment by diml** [comment]
I changed a bit the way the various functions call each other, and changed which functions were memoized, so this might affect the order in which we walk the graph
